jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Utf8Appendable.java

```diff
@@ -243,94 +302,187 @@ public void appendByte(byte b) throws IOException
+                    _state = current;
+                }
             }
         }
     }
 
+    public void appendReplacement()
+    {
+        try
+        {
+            _appendable.append(REPLACEMENT);
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeException(e);
+        }
+        _hasReplacements = true;
+    }
+
     public boolean isUtf8SequenceComplete()
     {
         return _state == UTF8_ACCEPT;
     }
 
-    @SuppressWarnings("serial")
-    public static class NotUtf8Exception extends IllegalArgumentException
+    /**
+     * Finish the buffer state.
+     * <p>
+     *     This will address any incomplete utf-8 sequences in the buffer, and set Utf8Appendable in a state
+     *     where it will be ready to accept new utf-8 sequences.
+     * </p>
+     */
+    public void finish()
     {
-        public NotUtf8Exception(String reason)
+        if (!isUtf8SequenceComplete())
         {
-            super("Not valid UTF8! " + reason);
+            appendReplacement();
         }
+        _codep = 0;
+        _state = UTF8_ACCEPT;
     }
 
-    public void checkState()
+    /**
+     * Clear out the {@link Appendable} of any content, resetting it to zero.
+     * <p>
+     *     This will not clear out any partial code points.
+     * </p>
+     */
+    public abstract void clear();
+
+    /**
+     * True if there was a replacement of a bad UTF-8 sequence with a UTF-8 replacement character.
+     *
+     * <p>
+     *     If the input already has a replacement character, this is not counted when using
+     *     this method.
+     * </p>
+     *
+     * @return true if there was a replacement of an appended character or byte with a replacement character.
+     */
+    public boolean hasReplacements()
     {
-        if (!isUtf8SequenceComplete())
-        {
-            _codep = 0;
-            _state = UTF8_ACCEPT;
-            try
-            {
-                _appendable.append(REPLACEMENT);
-            }
-            catch (IOException e)
-            {
-                throw new RuntimeException(e);
-            }
-            throw new NotUtf8Exception("incomplete UTF8 sequence");
-        }
+        return _hasReplacements;
     }
 
     /**
-     * @return The UTF8 so far decoded, ignoring partial code points
+     * Get the String.
+     *
+     * @param completeCodePoints true to complete any active codepoints/sequences before returning the String
+     * @param throwableOnReplacementsSupplier if a replacement of a character occurred during append,
+     * and if this supplier is non-null, then use this supplier to throw an error
+     * @return the String
+     * @throws X if replacement of bad characters occurred, and {@code throwableOnReplacementsSupplier} was provided
      */
-    public abstract String getPartialString();
+    public <X extends Throwable> String getString(boolean completeCodePoints, Supplier<? extends X> throwableOnReplacementsSupplier) throws X
```

Comment:
I don't love this API.  We already have some methods throwing `CharacterCodingException` and we have a convenience supplier for `NotUtf8Exception`.  Do we really want to open this class up to throw arbitrary other exceptions from this method?

If feels way too flexible for no good reason, without actually being flexible enough... e.g.  why a `Supplier`?  Maybe it should be a `Function` taking the bad code point and returning an `Exception`?

I'm back to worrying that we are making too big a change to this core class.   Let me ponder on this for a little while....

Reply:
This is the same signature as used in Optional.
We now have things throwing `BadPayloadException` in websocket, where it used to throw + catch + wrap + throw.
And we have `CharacterCodingException` flows as well.
This also eliminated a few other throw NotUtf8Exception + catch + wrap + throw different Exception scenarios.

Reply:
We don't know the bad code point at the time of `.getString()`, so a Function has no meaning.
If we start to track them (which I don't like as an idea), then we would also possible have more than 1 bad code point, then what?